pinot-minion/src/main/java/com/linkedin/pinot/minion/executor/SegmentConversionExecutor.java

```diff
@@ -0,0 +1,121 @@
+
+    File tempDataDir = new File(new File(MINION_CONTEXT.getDataDir(), taskType), "tmp-" + System.nanoTime());
+    Preconditions.checkState(tempDataDir.mkdirs());
+    try {
+      // Download the tarred segment file
+      File tarredSegmentFile = new File(tempDataDir, "tarredSegmentFile");
+      SegmentFetcherFactory.getSegmentFetcherBasedOnURI(downloadURL)
+          .fetchSegmentToLocal(downloadURL, tarredSegmentFile);
+
+      // Un-tar the segment file
+      File segmentDir = new File(tempDataDir, "segmentDir");
+      TarGzCompressionUtils.unTar(tarredSegmentFile, segmentDir);
+      File[] files = segmentDir.listFiles();
+      Preconditions.checkState(files != null && files.length == 1);
+      File indexDir = files[0];
+
+      // Convert the segment
+      File convertedSegmentDir = new File(tempDataDir, "convertedSegmentDir");
+      Preconditions.checkState(convertedSegmentDir.mkdir());
+      File convertedIndexDir = new File(convertedSegmentDir, segmentName);
+      convert(pinotTaskConfig, indexDir, convertedIndexDir);
+
+      // Tar the converted segment
+      File convertedTarredSegmentDir = new File(tempDataDir, "convertedTarredSegmentDir");
+      Preconditions.checkState(convertedTarredSegmentDir.mkdir());
+      File convertedTarredSegmentFile = new File(
+          TarGzCompressionUtils.createTarGzOfDirectory(convertedIndexDir.getPath(),
+              new File(convertedTarredSegmentDir, segmentName).getPath()));
+
+      // Check whether the task get cancelled before uploading the segment
+      if (_cancelled) {
+        throw new TaskCancelledException(
+            MinionConstants.ConvertToRawIndexTask.TASK_TYPE + " task on table: " + tableName + ", segment: "
+                + segmentName + " has been cancelled");
+      }
+
+      // Upload the converted tarred segment file with original segment crc and task type encoded in the http headers.
+      // Original segment crc is used to check whether the original segment get refreshed, so that the newer segment
+      // won't get override. We can decide how to handle the segment based on the task type on controller side.
+      // NOTE: even segment is not changed, still need to upload the segment to update the segment ZK metadata so that
+      // segment will not be submitted again
+      Header ifMatchHeader = new Header(HttpHeaders.IF_MATCH, originalSegmentCrc);
+      Header taskTypeHeader =
+          new Header(HttpHeaders.USER_AGENT, CommonConstants.Minion.HTTP_TASK_TYPE_HEADER_PREFIX + taskType);
+      List<Header> httpHeaders = Arrays.asList(ifMatchHeader, taskTypeHeader);
+      FileUploadUtils.sendFile(uploadURL, convertedSegmentDir.getName(),
+          new FileInputStream(convertedTarredSegmentFile), convertedTarredSegmentFile.length(),
+          FileUploadUtils.SendFileMethod.POST, httpHeaders);
+
+      LOGGER.info("Done executing ConvertToRawIndexTask on table: {}, segment: {}", tableName, segmentName);
```

Comment:
Any metrics to be updated regarding time taken to process a segment?

Reply:
Yes, in TaskFactoryRegistry